File: js/game/session.js
// One play-through of a level: moves, undo history, hearts, combos, scoring, hints.
// DOM-free so it can be unit tested.
import { buildStatic, initialState, remaining } from '../core/model.js';
import { applyMove, isStuck, legalMoves } from '../core/rules.js';
import { solve } from '../core/solver.js';

export const COMBO_WINDOW_MS = 2600;
export const MAX_MULT = 5;
const BASE_POINTS = 100;
const CHAIN_POINTS = 150;

export class Session {
  constructor(level, { mode = 'challenge', maxHearts = 3 } = {}) {
    this.level = level;
    this.S = buildStatic(level);
    this.st = initialState(level);
    this.mode = mode;
    this.maxHearts = maxHearts;
    this.hearts = maxHearts;
    this.history = [];
    this.score = 0;
    this.combo = 0;
    this.mult = 1;
    this.lastClearAt = -Infinity;
    this.mistakes = 0;
    this.hintsUsed = 0;
    this.undos = 0;
    this.continues = 0;
    this.maxMult = 1;
    this.maxChain = 0;
    this.moves = 0;
    this.status = 'playing'; // playing | won | failed
  }

  get total() {
    return this.level.arrows.length;
  }

  get left() {
    return remaining(this.st);
  }

  tap(i, now = Date.now()) {
    if (this.status !== 'playing') return { result: 'invalid', events: [] };
    const r = applyMove(this.level, this.S, this.st, i);
    if (r.result === 'blocked') {
      this.mistakes++;
      this.combo = 0;
      this.mult = 1;
      if (this.mode === 'challenge') {
        this.hearts = Math.max(0, this.hearts - 1);
        if (this.hearts === 0) this.status = 'failed';
      }
      return { ...r, heartsLeft: this.hearts };
    }
    if (r.result !== 'cleared') return r;

    this.history.push({ st: this.st, score: this.score });
    this.st = r.st;
    this.moves++;
    this.combo = now - this.lastClearAt <= COMBO_WINDOW_MS ? this.combo + 1 : 1;
    this.lastClearAt = now;
    const prevMult = this.mult;
    this.mult = Math.min(MAX_MULT, 1 + Math.floor((this.combo - 1) / 3));
    this.maxMult = Math.max(this.maxMult, this.mult);
    this.maxChain = Math.max(this.maxChain, r.chain);

    let gained = BASE_POINTS * this.mult;
    for (const e of r.events) if (e.type === 'clear' && e.chain > 0) gained += CHAIN_POINTS * e.chain * this.mult;
    this.score += gained;
    if (remaining(this.st) === 0) this.status = 'won';
    return { ...r, gained, mult: this.mult, multUp: this.mult > prevMult };
  }

  canUndo() {
    return this.history.length > 0 && this.status === 'playing';
  }

  undo() {
    if (!this.canUndo()) return false;
    const h = this.history.pop();
    this.st = h.st;
    this.score = h.score;
    this.combo = 0;
    this.mult = 1;
    this.lastClearAt = -Infinity;
    this.undos++;
    return true;
  }

  /** Next move toward a solution: arrow index, 'dead' if unsolvable from here, or null if unknown. */
  findHint() {
    const r = solve(this.level, this.S, this.st, { maxNodes: 30000 });
    if (r.solved) return r.moves[0];
    if (r.exhausted) {
      const legal = legalMoves(this.level, this.S, this.st);
      return legal.length ? legal[0] : 'dead';
    }
    return 'dead';
  }

  useHint() {
    this.hintsUsed++;
  }

  stuck() {
    return this.status === 'playing' && isStuck(this.level, this.S, this.st);
  }

  continueRun(hearts = 2) {
    this.hearts = hearts;
    this.status = 'playing';
    this.continues++;
  }

  results() {
    const perfect = this.mistakes === 0 && this.hintsUsed === 0;
    const slips = this.mistakes + this.hintsUsed;
    const stars = perfect ? 3 : slips <= 1 ? 2 : 1;
    const perfectBonus = perfect ? 500 + 25 * this.total : 0;
    const heartBonus = this.mode === 'challenge' ? this.hearts * 100 : 0;
    return {
      stars,
      perfect,
      base: this.score,
      perfectBonus,
      heartBonus,
      total: this.score + perfectBonus + heartBonus,
      maxMult: this.maxMult,
      maxChain: this.maxChain,
    };
  }
}

File: mobile/lib/game/session.dart
// One play-through of a level: moves, undo history, hearts, combos, scoring, hints.
// Port of js/game/session.js (UI-free, unit tested).
import 'dart:math' as math;

import '../core/model.dart';
import '../core/rules.dart';
import '../core/solver.dart';

const int comboWindowMs = 2600;
const int maxMult = 5;
const int _basePoints = 100;
const int _chainPoints = 150;

enum GameMode { challenge, zen }

enum SessionStatus { playing, won, failed }

class TapOutcome {
  TapOutcome(this.result, {this.move, this.gained = 0, this.mult = 1, this.multUp = false, this.heartsLeft});
  final MoveResult result;
  final MoveOutcome? move;
  final int gained;
  final int mult;
  final bool multUp;
  final int? heartsLeft;

  List<GameEvent> get events => move?.events ?? const [];
  Ray? get trace => move?.trace;
  int get chain => move?.chain ?? 0;
}

class Results {
  Results(this.stars, this.perfect, this.base, this.perfectBonus, this.heartBonus, this.maxMult, this.maxChain);
  final int stars;
  final bool perfect;
  final int base, perfectBonus, heartBonus, maxMult, maxChain;
  int get total => base + perfectBonus + heartBonus;

  Map<String, dynamic> toJson() => {
        'stars': stars,
        'perfect': perfect,
        'base': base,
        'perfectBonus': perfectBonus,
        'heartBonus': heartBonus,
        'total': total,
        'maxMult': maxMult,
        'maxChain': maxChain,
      };
}

/// Hint answer: a move index, or [HintResult.dead] when unsolvable from here.
class HintResult {
  const HintResult._(this.move);
  const HintResult.move(int m) : this._(m);
  static const dead = HintResult._(null);
  final int? move;
  bool get isDead => move == null;
}

class Session {
  Session(this.level, {this.mode = GameMode.challenge, this.maxHearts = 3})
      : s = buildStatic(level),
        st = initialState(level),
        hearts = maxHearts;

  final Level level;
  final StaticBoard s;
  PlayState st;
  final GameMode mode;
  final int maxHearts;
  int hearts;
  final List<(PlayState, int)> _history = [];
  int score = 0;
  int combo = 0;
  int mult = 1;
  int _lastClearAt = -1 << 40;
  int mistakes = 0;
  int hintsUsed = 0;
  int undos = 0;
  int continues = 0;
  int maxMultReached = 1;
  int maxChain = 0;
  int moves = 0;
  SessionStatus status = SessionStatus.playing;
  bool statsRecorded = false;

  int get total => level.arrows.length;
  int get left => st.remaining;

  TapOutcome tap(int i, int now) {
    if (status != SessionStatus.playing) return TapOutcome(MoveResult.invalid);
    final r = applyMove(level, s, st, i);
    if (r.result == MoveResult.blocked) {
      mistakes++;
      combo = 0;
      mult = 1;
      if (mode == GameMode.challenge) {
        hearts = math.max(0, hearts - 1);
        if (hearts == 0) status = SessionStatus.failed;
      }
      return TapOutcome(MoveResult.blocked, move: r, heartsLeft: hearts);
    }
    if (r.result != MoveResult.cleared) return TapOutcome(r.result, move: r);

    _history.add((st, score));
    st = r.st;
    moves++;
    combo = now - _lastClearAt <= comboWindowMs ? combo + 1 : 1;
    _lastClearAt = now;
    final prevMult = mult;
    mult = math.min(maxMult, 1 + (combo - 1) ~/ 3);
    maxMultReached = math.max(maxMultReached, mult);
    maxChain = math.max(maxChain, r.chain);

    var gained = _basePoints * mult;
    for (final e in r.events) {
      if (e is ClearEvent && e.chain > 0) gained += _chainPoints * e.chain * mult;
    }
    score += gained;
    if (st.remaining == 0) status = SessionStatus.won;
    return TapOutcome(MoveResult.cleared, move: r, gained: gained, mult: mult, multUp: mult > prevMult);
  }

  bool get canUndo => _history.isNotEmpty && status == SessionStatus.playing;

  bool undo() {
    if (!canUndo) return false;
    final (prev, prevScore) = _history.removeLast();
    st = prev;
    score = prevScore;
    combo = 0;
    mult = 1;
    _lastClearAt = -1 << 40;
    undos++;
    return true;
  }

  HintResult findHint() {
    final r = solve(level, s, st, maxNodes: 30000);
    if (r.solved) return HintResult.move(r.moves.first);
    if (r.exhausted) {
      final legal = legalMoves(level, s, st);
      return legal.isNotEmpty ? HintResult.move(legal.first) : HintResult.dead;
    }
    return HintResult.dead;
  }

  void useHint() => hintsUsed++;

  bool get stuck => status == SessionStatus.playing && isStuck(level, s, st);

  void continueRun([int newHearts = 2]) {
    hearts = newHearts;
    status = SessionStatus.playing;
    continues++;
  }

  Results results() {
    final perfect = mistakes == 0 && hintsUsed == 0;
    final slips = mistakes + hintsUsed;
    final stars = perfect ? 3 : (slips <= 1 ? 2 : 1);
    return Results(
      stars,
      perfect,
      score,
      perfect ? 500 + 25 * total : 0,
      mode == GameMode.challenge ? hearts * 100 : 0,
      maxMultReached,
      maxChain,
    );
  }
}
